### DISCUSSION TITLE
What is the role of the downsample_rate variable？

### DISCUSSION THREAD

**Discussion by YangHao97** [s3prl/s3prl#411] [Q&A]
Hi, I'm adding my encoder into upstream models, and I set downsample_rate = 320. But as the following code block, it can not run correctly.
![image](https://user-images.githubusercontent.com/71253402/194845112-3c0e8482-e46f-4f09-82cf-5c5bcd494ff2.png)
The length of my output feature is about 1400-2000, it longer than feature_len = round(len(wav) / self.downsample_rate (length = 200-300). When I want the downsample rate can be calculated dynamically, I found it seems the SUPERB codes are not completed.
Q1: Why do we need a downsample rate?
Q2: Am I missing the right way to use dynamic downsample rate?
Thank you!

**Comment by leo19941227** [comment]
Hi,

Can you give an example of the input waveform lengths v.s. output feature lengths of your encoder?

For Q1, the downsample rate property is useful to prepare training labels when we want frame level classification/regression. Like speaker diarization, speech enhancement and separation.

**Comment by YangHao97** [comment]
Hi, the output feature length is fixed. For each sample input, we always get the feature with the same length. I mentioned (1400-2000) is because we have several different encoders, they have different output length. But for any one encoder, the output length is fixed.

**Comment by bagustris** [comment]
I think this topic should be moved into the discussion (instead of the issue) and the (new) documentation should clarify the role of `donwnsample_rate`, including how to use this variable on new data (an example). Many people asked about this, including me.

**Reply by leo19941227** [reply]
Sounds good!

**Comment by leo19941227** [comment]
> Hi, the output feature length is fixed. For each sample input, we always get the feature with the same length. I mentioned (1400-2000) is because we have several different encoders, they have different output length. But for any one encoder, the output length is fixed.

You mean, no matter it is a 3 secs utterance or a 20 secs utterance, your encoder always produce the same output sequence length? Then this representation will not be suitable for tasks requiring variable length representation like ASR and diarization right?
If so, I think you might want to do a simple trick to do sliding window when extracting representation, like feeding 1 sec window with 0.5 sec overlap  into your encoder, so you can have a variable length of sequence of representation.

**Reply by YangHao97** [reply]
Thanks for you help, I'll try it.
